### DISCUSSION TITLE
int ve char değişkenlerini main'in dışında tanımlamanın yararı nedir?

### DISCUSSION THREAD

**Discussion by baranar** [19/classroom#60] [C] [answered]
başlıkta da olduğu gibi:

```c
int sayi1, sayi2, sonuc;
char islem;
#include <stdio.h>
int main()
{
    printf("Ufak capli hesap makinesi\n");
    printf("*************************\n\n\n");
    printf("Lutfen ilk sayiyi giriniz: ");
    scanf("%d",&sayi1);
    printf("Lutfen ikinci sayiyi giriniz: ");
    scanf("%d",&sayi2);
    printf("lutfen isleminizi giriniz: ");
    scanf("%s",&islem);
    switch(islem)
    {
        case '+': 
        sonuc = sayi1 + sayi2;
        printf("isleminizin sonucu: %d", sonuc);
        break;
    }
}
```

bu şekilde tanımladığımda kod sorunsuz çalışırken;

```c
#include <stdio.h>
int main()
{
    int sayi1, sayi2, sonuc;
    char islem;
    printf("Ufak capli hesap makinesi\n");
    printf("*************************\n\n\n");
    printf("Lutfen ilk sayiyi giriniz: ");
    scanf("%d",&sayi1);
    printf("Lutfen ikinci sayiyi giriniz: ");
    scanf("%d",&sayi2);
    printf("lutfen isleminizi giriniz: ");
    scanf("%s",&islem);
    switch(islem)
    {
        case '+': 
        sonuc = sayi1 + sayi2;
        printf("isleminizin sonucu: %d", sonuc);
        break;
    }
}
```

int ve char'ı bu şekilde tanımlarsam mesela 10+20 işleminin çıktısını bana 10 olarak veriyor.
Bunun nedeni nedir? daha pratik çözümü var mıdır?

**Comment by roktas** [comment]
Bu soru C dersinde izlediğimiz **ajanda**da "kapsam" ("scope") konusuna gelindiğinde cevaplayabileceğin "erken" bir soru.  O tarihe kadar burada kalsın.  Konuyu işlediğimizde hala yanıt bulamamışsan hatırlat bu soru özelinde ayrıntılı bir cevap vereyim.

P.S. Soru metnini biraz biçimledim.  Kod örneklerini "fenced code block"larla yazın.  Ayrıntılı bilgi için bk. [Github Markdown](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax)

**Reply by baranar** [reply]
Teşekkürler hocam

**Comment by roktas** [comment, marked as answer]
@baranar 30 Mart'ta yaptığım derste bu sorunla ilgili ipucu verdim.  Daha açık yazayım, aşağıdaki satırlar üzerinde düşün:

```c
char islem;

...

printf("lutfen isleminizi giriniz: ");
scanf("%s",&islem);
```

**Reply by baranar** [reply]
karakter alacağım için %s 'yi %c ile değiştirdim. Daha sonra tam sayıları aldığım ve karakteri aldığım scanf'lerin arasına getchar() kullandığımda program çalıştı. Doğru adımlar mı? diğer sorum: Programın ilk halinde değişkenleri main klasörünün dışında tanımladığımda çalışıyordu. mainin içinde ya da dışında tanımlamanın farkı nedir?

**Reply by roktas** [reply]
Buradaki temel sorun karakter okurken `%c` yerine `%s` kullanman idi.  Bu tür hatalı kullanımlarda `scanf`'in davranışı belirsiz olur.  Konunun kapsam (scope) ile olan ilişkisi de bu belirsizlik ile ilişkili.  Bugün böyle davranır, yarın başka bir gerçeklemede veya başka bir platformda farklı davranabilir.  Bu bir ayrıntı, üzerinde durman gerekmiyor.  Temel soruna odaklan.

P.S. `scanf`'le başlamışsan onunla devam et, ayrıca `getchar` kullanmana gerek yok; bu konuyu dağıtmaktan başka bir işe yaramaz.
